### PR TITLE
fix(Android|pip): do not invoke 'enterPictureInPicture' in PAUSED state

### PR DESCRIPTION
**Pull request by paweldomas**
Activity.enterPictureInPictureMode method must be invoked synchronously
on userLeaveHint callback in order to be sure that the current Activity
is still visible (does not transit to PAUSED state). Previously if the
asynchronous processing would be delayed enough for the Activity to go
into the PAUSED state it will be too late to go into the PiP mode.

### CODE REVIEW

**Comment by paweldomas** [comment]
This is a proposal and some docs are not complete. Plus I don't know how to deal with the lack of constants for the external API event names.

File: android/sdk/src/main/java/org/jitsi/meet/sdk/ExternalAPIModule.java [hunk 1]
```diff
@@ -130,6 +130,12 @@ public void sendEvent(String name, ReadableMap data, String scope) {
             return;
         }
 
+        if (name.equals("CONFERENCE_WILL_JOIN")) {
+            view.setCurrentConferenceUrl(data.getString("url"));
+        } else if (name.equals("CONFERENCE_WILL_LEAVE")) {
+            view.setCurrentConferenceUrl(null);
```
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/ExternalAPIModule.java]: how about failure? also, if we follow a link while in a conference, do we always get ended / failed for the ongoing conference **before** will join from the new one?

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetView.java [hunk 2]
```diff
@@ -498,6 +486,37 @@ public void loadURLString(@Nullable String urlString) {
         loadURLObject(urlObject);
     }
 
+    void setCurrentConferenceUrl(String conferenceUrl) {
+        System.err.println(hashCode()+ " EXT CONF URL: " + conferenceUrl);
```
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetView.java]: debug leftover?

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetView.java [hunk 3]
```diff
@@ -257,37 +275,7 @@ public static void onNewIntent(Intent intent) {
...
-                = reactInstanceManager.getCurrentReactContext();
-            if (reactContext != null) {
-                reactContext
-                    .getJSModule(
-                        DeviceEventManagerModule.RCTDeviceEventEmitter.class)
-                    .emit(eventName, params);
-            }
-        }
-    }
+    private String conferenceUrl;
```
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetView.java]: does this not need to be synchronized?

File: android/sdk/src/main/java/org/jitsi/meet/sdk/ExternalAPIModule.java (hunk 1 above)
Comment (paweldomas) [reply android/sdk/src/main/java/org/jitsi/meet/sdk/ExternalAPIModule.java]: Good point ! I need to check all that

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetView.java (hunk 3 above)
Comment (paweldomas) [reply android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetView.java]: I don't know... currently only one thread writes there and one reads, so maybe volatile keyword would be enough

**Comment by paweldomas** [comment]
> should the current conference URL have a public setter? if so, should we also have it on iOS?

I don't know about that. It will not be used there currently I guess, so what's the point ?

**Comment by paweldomas** [comment]
@saghul what will happen with this logic here https://github.com/jitsi/jitsi-meet/blob/master/react/features/mobile/picture-in-picture/actions.js#L32 ? we'll have two logics for "is ok to enter PiP"

**Comment by saghul** [comment]
@paweldomas That's a good point. So, the thing here is that the PiP feature currently has 2 entry-points:

- the onUserLeaveHit activity life-cycle method
- the PiP button

One starts from native one from JS, so AFAICT, we need the checks on both places to be extra sure. If you have a suggestion on how to simplify this, I'm listening :-)

**Comment by paweldomas** [comment]
So one way which I do not like would be to eventually swallow the "enter PiP mode" external API event on the native side, somewhere around where it sets the URL in this PR. Because the URL seems to be "is OK to enter PiP" flag. The problem with that we'll be doing the same on iOS - I'm not sure how big of a problem that would be.

My preferred way is to go with my original PR which extracts isOkToEnterPiP function. I just am still not convinced neither by you nor @lmarinov that it can be simplified with the "set room" or "conference URL" on the native part without duplication.

**Comment by paweldomas** [comment]
@saghul now I recall you were proposing to not have any check there given that the action can come only from the button displayed on the conference view. I don't feel it's fine to do it this way, but here's a commit which does that

**Comment by paweldomas** [comment]
@saghul rebased and force pushed with minor improvement and filled missing docs

**Comment by saghul** [comment]
@paweldomas sorry man, looks like the mobile-invite thing caused more conflicts 😢  can you please fix it? (last time, I promise!)

**Comment by paweldomas** [comment]
@saghul yes, I'm doing it right now

**Comment by saghul** [comment]
Note: I merged this with a (confirmed) unrelated failing test on Safari.

**Pull request merged**